src/os/bluestore/Btree2Allocator.h

```diff
@@ -0,0 +1,425 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <mutex>
+
+#include "include/cpp-btree/btree_map.h"
+#include "include/cpp-btree/btree_set.h"
+
+#include "Allocator.h"
+
+#include "os/bluestore/bluestore_types.h"
+#include "include/mempool.h"
+#include "common/ceph_mutex.h"
+
+template <size_t BASE_BITS, size_t NUM_BUCKETS, size_t MUX = 1>
+struct SizeBucketCollectionBase {
+  static const size_t base_bits = BASE_BITS;
+  static const size_t base = 1ull << base_bits;
+  static const size_t mux = MUX;
+  static const size_t num_size_buckets = NUM_BUCKETS;
+  static const size_t bucket_max = base << ((mux * (num_size_buckets - 2)));
+
+  inline size_t _get_size_bucket(uint64_t len) const {
+    size_t idx;
+    if (len <= base) {
+      idx = 0;
+    } else if (len > bucket_max) {
+      idx = num_size_buckets - 1;
+    } else {
+      size_t most_bit = cbits(uint64_t(len - 1)) - 1;
+      idx = 1 + ((most_bit - base_bits) / mux);
+    }
+    ceph_assert(idx < num_size_buckets);
+    return idx;
+  }
+};
+
+class ChunkCache {
+public:
+  class LocklessBuf {
```

Comment:
Lockless optional buffer is unusual concept.
Some description on its role and scope of its applicability would be nice.

Reply:
Renamed to LocklessWillfulStack and moved to Allocator.h